overlord/devicestate/firstboot_preseed_test.go

```diff
@@ -426,3 +430,214 @@ snaps:
+	snapYaml = `name: bar
+base: core18
+version: 1.0
+slots:
+ shared-data-slot:
+  interface: content
+  content: mylib
+  read:
+   - /
+apps:
+ bar:
+  command: bin/bar
+`
+	snapFiles := [][]string{
+		{"bin/bar", ``},
+	}
+	barFname, barDecl, barRev := s.MakeAssertedSnap(c, snapYaml, snapFiles, snap.R(65), "developerid")
+	s.WriteAssertions("bar.asserts", s.devAcct, barDecl, barRev)
+
+	// add a model assertion and its chain
+	assertsChain := s.makeModelAssertionChain(c, "my-model-classic", nil)
+	s.WriteAssertions("model.asserts", assertsChain...)
+
+	// create a seed.yaml
+	content := []byte(fmt.Sprintf(`
+snaps:
+ - name: snapd
+   file: %s
+ - name: core18
+   file: %s
+ - name: foo
+   file: %s
+ - name: bar
+   file: %s
+`, snapdFname, core18Fname, fooFname, barFname))
+	err := ioutil.WriteFile(filepath.Join(dirs.SnapSeedDir, "seed.yaml"), content, 0644)
+	c.Assert(err, IsNil)
+
+	// run the firstboot stuff
+	s.startOverlord(c)
+	st := s.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	opts := &devicestate.PopulateStateFromSeedOptions{Preseed: true}
+	tsAll, err := devicestate.PopulateStateFromSeedImpl(st, opts, s.perfTimings)
+	c.Assert(err, IsNil)
+	// use the expected kind otherwise settle with start another one
+	chg := st.NewChange("seed", "run the populate from seed changes")
+	for _, ts := range tsAll {
+		chg.AddAll(ts)
+	}
+	c.Assert(st.Changes(), HasLen, 1)
+
+	checkPreseedOrder(c, tsAll, "snapd", "core18", "foo", "bar")
+
+	st.Unlock()
+	err = s.overlord.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Assert(chg.Err(), IsNil)
+	c.Check(chg.Status(), Equals, state.DoingStatus)
+
+	// we are done with this instance of the overlord, stop it here. Otherwise
+	// it will interfere with our second overlord instance
+	c.Assert(s.overlord.Stop(), IsNil)
+
+	// Verify state between the two change runs
+	r, err := os.Open(dirs.SnapStateFile)
+	c.Assert(err, IsNil)
+	diskState, err := state.ReadState(nil, r)
+	c.Assert(err, IsNil)
+
+	diskState.Lock()
+	defer diskState.Unlock()
+
+	// seeded snaps are installed
+	_, err = snapstate.CurrentInfo(diskState, "snapd")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "core18")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "foo")
+	c.Check(err, IsNil)
+	_, err = snapstate.CurrentInfo(diskState, "bar")
+	c.Check(err, IsNil)
+
+	// but we're not considered seeded
+	var seeded bool
+	err = diskState.Get("seeded", &seeded)
+	c.Assert(err, testutil.ErrorIs, state.ErrNoState)
+
+	// For the next step of the test, we want to turn off pre-seeding so
+	// we can run the change all the way through, and also see the hooks go
+	// off.
+	restore = snapdenv.MockPreseeding(false)
+	defer restore()
+
+	// Create a new overlord after turning pre-seeding off. We don't want to
+	// overwrite the old s.overlord variable as this will interfere with the
+	// deferred cleanup, so we do it ourselves here and keep it separate.
+	o, err := overlord.New(nil)
```

Comment:
I think the intention was that startOverlord could be used for this situation, I think with these changes it should work:

```
diff --git a/overlord/devicestate/firstboot_preseed_test.go b/overlord/devicestate/firstboot_preseed_test.go
index d8fb663496..2aa6ba1118 100644
--- a/overlord/devicestate/firstboot_preseed_test.go
+++ b/overlord/devicestate/firstboot_preseed_test.go
@@ -31,7 +31,6 @@ import (
 
        "github.com/snapcore/snapd/dirs"
        "github.com/snapcore/snapd/interfaces"
-       "github.com/snapcore/snapd/overlord"
        "github.com/snapcore/snapd/overlord/devicestate"
        "github.com/snapcore/snapd/overlord/hookstate"
        "github.com/snapcore/snapd/overlord/restart"
@@ -575,11 +574,9 @@ snaps:
        // Create a new overlord after turning pre-seeding off. We don't want to
        // overwrite the old s.overlord variable as this will interfere with the
        // deferred cleanup, so we do it ourselves here and keep it separate.
-       o, err := overlord.New(nil)
-       c.Assert(err, IsNil)
-       o.StartUp()
+       s.startOverlord(c)
+       st = s.overlord.State()
 
-       st = o.State()
        st.Lock()
        defer st.Unlock()
 
@@ -588,16 +585,16 @@ snaps:
        chg1.SetStatus(state.DoingStatus)
 
        st.Unlock()
-       err = o.Settle(settleTimeout)
+       err = s.overlord.Settle(settleTimeout)
        st.Lock()
        c.Assert(err, IsNil)
 
        restart.MockPending(st, restart.RestartUnset)
        st.Unlock()
-       err = o.Settle(settleTimeout)
+       err = s.overlord.Settle(settleTimeout)
        st.Lock()
        c.Assert(err, IsNil)
-       c.Assert(o.Stop(), IsNil)
+       c.Assert(s.overlord.Stop(), IsNil)
        c.Assert(err, IsNil)
 
        // Update the change pointer to the change in the new state
@@ -624,20 +621,4 @@ snaps:
        var conns map[string]interface{}
        c.Assert(st.Get("conns", &conns), IsNil)
        c.Assert(conns, HasLen, 2)
-
-       repo := o.InterfaceManager().Repository()
-       cn, err := repo.Connected("foo", "shared-data-plug")
-       c.Assert(err, IsNil)
-       c.Assert(cn, HasLen, 1)
-       c.Assert(cn, DeepEquals, []*interfaces.ConnRef{{
-               PlugRef: interfaces.PlugRef{Snap: "foo", Name: "shared-data-plug"},
-               SlotRef: interfaces.SlotRef{Snap: "bar", Name: "shared-data-slot"},
-       }})
-       cn, err = repo.Connected("foo", "network")
-       c.Assert(err, IsNil)
-       c.Ass
-
diff --git a/overlord/devicestate/firstboot_test.go b/overlord/devicestate/firstboot_test.go
index 8b875ea3ae..a7ab7a6a4c 100644
--- a/overlord/devicestate/firstboot_test.go
+++ b/overlord/devicestate/firstboot_test.go
@@ -143,6 +143,9 @@ func (t *firstBootBaseTest) startOverlord(c *C) {
        t.overlord = ovld
        t.AddCleanup(func() {
                devicestate.EarlyConfig = nil
+               if t.overlord == nil {
+                       return
+               }
                t.overlord.Stop()
                t.overlord = nil
        })
```

Reply:
the comments would need to change, as now we deal fine with the overlord overwrite, as anyway we were calling Stop ourselves

Reply:
Ah nice - I was afraid I was ignoring some cleanup doing this, so I created it manually to ensure I didn't. Thanks for this suggestion!